src/crimson/os/seastore/transaction_manager.cc

```diff
@@ -569,64 +569,67 @@ TransactionManager::get_extent_if_live_ret TransactionManager::get_extent_if_liv
 
   return cache->get_extent_if_cached(t, addr, type
   ).si_then([this, FNAME, &t, type, addr, laddr, len](auto extent)
-	    -> get_extent_if_live_ret {
-    if (extent) {
+	    -> get_extents_if_live_ret {
+    if (extent && extent->get_length() == (extent_len_t)len) {
       DEBUGT("{} {}~{} {} is live in cache -- {}",
              t, type, laddr, len, addr, *extent);
-      return get_extent_if_live_ret (
+      std::list<CachedExtentRef> res;
+      res.emplace_back(std::move(extent));
+      return get_extents_if_live_ret(
 	interruptible::ready_future_marker{},
-	extent);
+	res);
     }
 
     if (is_logical_type(type)) {
       using inner_ret = LBAManager::get_mapping_iertr::future<CachedExtentRef>;
-      return lba_manager->get_mapping(
+      return lba_manager->get_mappings(
 	t,
-	laddr).si_then([=, &t] (LBAPinRef pin) -> inner_ret {
-	  ceph_assert(pin->get_key() == laddr);
-	  if (pin->get_val() == addr) {
-	    if (pin->get_length() != (extent_len_t)len) {
-	      ERRORT(
-		"Invalid pin {}~{} {} found for "
-		"extent {} {}~{} {}",
-		t,
-		pin->get_key(),
-		pin->get_length(),
-		pin->get_val(),
-		type,
-		laddr,
-		len,
-		addr);
-	      ceph_abort();
-	    }
-	    return cache->get_extent_by_type(
-	      t,
-	      type,
-	      addr,
-	      laddr,
-	      len,
-	      [this, pin=std::move(pin)](CachedExtent &extent) mutable {
-		auto lref = extent.cast<LogicalCachedExtent>();
-		assert(!lref->has_pin());
-		assert(!lref->has_been_invalidated());
-		assert(!pin->has_been_invalidated());
-		lref->set_pin(std::move(pin));
-		lba_manager->add_pin(lref->get_pin());
-	      }).si_then([=, &t](auto ret) {;
-		DEBUGT("{} {}~{} {} is live as logical extent -- {}",
-		       t, type, laddr, len, addr, extent);
-		return ret;
+	laddr,
+	len
+      ).si_then([=, &t](lba_pin_list_t pin_list) {
+	return seastar::do_with(
+	  std::list<CachedExtentRef>(),
+	  std::move(pin_list),
+	  [=, &t](std::list<CachedExtentRef> &list, lba_pin_list_t &pin_list) {
+	    auto &seg_addr = addr.as_seg_paddr();
+	    auto seg_addr_id = seg_addr.get_segment_id();
+	    return trans_intr::parallel_for_each(pin_list, [=, &seg_addr, &list, &t](LBAPinRef &pin) ->
+						 Cache::get_extent_iertr::future<> {
+	      auto pin_laddr = pin->get_key();
+	      auto pin_paddr = pin->get_val();
+	      auto pin_len = pin->get_length();
+
+	      auto &pin_seg_addr = pin_paddr.as_seg_paddr();
+	      auto pin_seg_addr_id = pin_seg_addr.get_segment_id();
+
+	      if (pin_seg_addr_id != seg_addr_id ||
+		  pin_paddr < seg_addr ||
+		  pin_paddr.add_offset(pin_len) > seg_addr.add_offset(len)) {
```

Comment:
See https://tracker.ceph.com/issues/66924 and https://github.com/ceph/ceph/pull/47553/commits/ec2c48343a6485ba03cb9bd87b3b0fb59379e402,

The issue proves that the extent has been moved to **the same segment** during reclaiming **this segment**, which looks strange because this segment should already be closed upon reclaiming, which means the segment is impossible to be modified upon reclaiming.

What do you think @xxhdx1985126 @zhscn ?

Reply:
@cyx1231st Which branch are you testing?

Reply:
> See https://tracker.ceph.com/issues/66924 and [ec2c483](https://github.com/ceph/ceph/commit/ec2c48343a6485ba03cb9bd87b3b0fb59379e402),
> 
> The issue proves that the extent has been moved to **the same segment** during reclaiming **this segment**, which looks strange because this segment should already be closed upon reclaiming, which means the segment is impossible to be modified upon reclaiming.
> 
> What do you think @xxhdx1985126 @zhscn ?

Does the test environment use multiple devices?

Reply:
Latest main + testing PRs (https://github.com/ceph/ceph/labels/wip-yingxin-testing).

I think the official rados-crimson suite won't use multiple devices for an OSD.

Reply:
Seems there are two IO transactions:
1. write data with offset 0 with length 1MB;
2. write data with offset 4K with length 4K, which caused lba mapping split;
Both two transactions write data to the same segement:
![drawio](https://github.com/user-attachments/assets/2b7444d3-c2c3-4dd4-9ad9-bb0b75c2ab6d)
The assertioin will fail in this case.

Reply:
Sorry forget to update the test log, added in the tracker.

Reply:
@zhscn I think so, seems the segment `Seg[Dev(0),5]` is just closed (04:56:50,177) before it is reclaming (04:56:50,183), so it's possible that a concurrent transaction is moving/splitting the extent into the same segment in lba, even though the segment containing the extent is closed.

Should be a simple fix, then.

Reply:
The problem could be worse than expected. The root cause of this issue is that GC transactions can't retrieve all backref entries. At T1, GC transaction fetch all backref entries located at to be reclaimed region, and after the fetch completes at T2, an IO transaction invokes `Cache::complete_commit` which updates the cached backref entries.

Enabling multiple devices in crimson will make this process more complicated.

Reply:
I see, it is a separate issue. I think a simple solution is to add another runtime state to the segments to indicate that all the transactions attempting to write a segment has been completed. And allow a segment to be considered for reclaim only after that.

Reply:
Created another https://tracker.ceph.com/issues/66941